### PR TITLE
scaffolder: Collect the users `oauth` token using `RepoUrlPicker` and allow templating of `secrets` in template inputs

### PR DESCRIPTION
**Pull request by benjdlambert**
This will be the last PR in a long list of stuff to get #5214 across the line. There's a little bit left to do to open up some of the actions to receive the token as a parameter, so the action doesn't assume the name of the secret and some more documentation and one more test. But opening this up so we can see it all so far.

### CODE REVIEW

File: plugins/scaffolder/src/components/fields/RepoUrlPicker/utils.ts [hunk 1]
```diff
@@ -44,22 +44,22 @@ export function serializeRepoPickerUrl(data: RepoUrlPickerState) {
 export function parseRepoPickerUrl(
   url: string | undefined,
 ): RepoUrlPickerState {
-  let host = undefined;
-  let owner = undefined;
-  let repoName = undefined;
-  let organization = undefined;
-  let workspace = undefined;
-  let project = undefined;
+  let host = '';
```
Comment (benjdlambert) [inline plugins/scaffolder/src/components/fields/RepoUrlPicker/utils.ts:47]: Had to change all these values from `undefined` to `''` as these values are passed to `mui` inputs and they are controlled components. An error is thrown if the component has a controlled value that moves from `undefined` to `''` so should follow something similar to this: https://reactjs.org/docs/forms.html#controlled-components

Comment (benjdlambert) [reply plugins/scaffolder/src/components/fields/RepoUrlPicker/utils.ts:47]:
```
Warning: ComponentName is changing an uncontrolled input of type text to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component
```

**Comment by benjdlambert** [comment]
Just another note here, realised that things like `fetch:template` could also need to be updated to use the users oauth token, but the backend will need to read the files anyways for importing however there could be template files located in places that the backend github app does not have access to.  Maybe we could skip this requirement for now and consider adding an option to `readUrl` and `readTree` that takes in additional headers and construct the `Authorization` header in the `fetch:template` and `fetch:plain` if a token is passed.

File: .changeset/twenty-queens-scream.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-scaffolder-backend': patch
+---
+
+Added support for templating secrets into actions input, and also added an extra `token` input argument to all publishers to provide a token that would override the `integrations.config`
```
Comment (jhaals) [inline .changeset/twenty-queens-scream.md:5]: should this perhaps also link out to the docs section that you updated in this PR?

**Review by jhaals — COMMENTED** [review]
Nice work Ben! 🧙 ⭐

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubActionsDispatch.ts [hunk 3]
```diff
@@ -68,18 +69,31 @@ export function createGithubActionsDispatchAction(options: {
               'Inputs keys and values to send to GitHub Action configured on the workflow file. The maximum number of properties is 10. ',
             type: 'object',
           },
+          token: {
+            title: 'Authentication Token',
+            type: 'string',
+            description: 'The GITHUB_TOKEN to use for authorization to GitHub',
```
Comment (Rugvip) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubActionsDispatch.ts:75]:
```suggestion
            description: 'The token to use for authorization to GitHub',
```

Bit worried about causing confusion around this being some kind of constant. Same for all of the other occurrences of this too

File: docs/features/software-templates/writing-templates.md [hunk 4]
```diff
@@ -287,6 +287,79 @@ The `RepoUrlPicker` is a custom field that we provide part of the
...
+    - title: Choose a location
+      required:
+        - repoUrl
+      properties:
+        repoUrl:
+          title: Repository Location
+          type: string
+          ui:field: RepoUrlPicker
+          ui:options:
+            # here's the new option you can pass to the RepoUrlPicker
```
Comment (Rugvip) [inline docs/features/software-templates/writing-templates.md:325]:
```suggestion
            # Here's the option you can pass to the RepoUrlPicker
```

Future-proofing the docs 👌

File: plugins/scaffolder/api-report.md [hunk 5]
```diff
@@ -164,6 +164,16 @@ export interface RepoUrlPickerUiOptions {
   allowedHosts?: string[];
   // (undocumented)
   allowedOwners?: string[];
+  // (undocumented)
+  requestUserCredentials?: {
+    resultSecretsKey: string;
```
Comment (Rugvip) [inline plugins/scaffolder/api-report.md:169]: Wondering if there's a better name here? I find it to be a bit unintuitive looking at it.

`secretsKey`
`secretsOutputKey`
`saveToSecretsKey`
`secretsSaveKey`
`secretsTemplateKey`
`templateSecretsKey`

File: plugins/scaffolder/api-report.md [hunk 6]
```diff
@@ -317,4 +327,9 @@ export const TextValuePicker: ({
   idSchema,
   placeholder,
 }: FieldProps<string>) => JSX.Element;
+
+// @public
+export const useSecretsContext: () => {
```
Comment (Rugvip) [inline plugins/scaffolder/api-report.md:332]: Wonder if we should make this slightly more verbose, and probably fine to skip the context suffix? Thinking something like

`useTemplateSecrets`
`useFieldExtensionSecrets`
`useTemplateFormSecrets`

**Review by Rugvip — COMMENTED** [review]
Awesome stuff :grin: 🎉 

Just a bunch of docs/naming suggestions

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubActionsDispatch.ts (hunk 3 above)
Comment (benjdlambert) [reply plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubActionsDispatch.ts:75]: Hmm yeah good point - I was just thinking that it seemed most familiar, but I get the confusion will update for all providers.

File: docs/features/software-templates/writing-templates.md (hunk 4 above)
Comment (benjdlambert) [reply docs/features/software-templates/writing-templates.md:325]: 👍

File: plugins/scaffolder/api-report.md (hunk 5 above)
Comment (benjdlambert) [reply plugins/scaffolder/api-report.md:169]: Yeah naming this thing has been hard, `secretsKey` is probably the most I like, but to be honest I don't LOVE all of them including the one I suggested.

Let's go with `secretsKey` for now. 👍

File: plugins/scaffolder/api-report.md (hunk 6 above)
Comment (benjdlambert) [reply plugins/scaffolder/api-report.md:332]: `useTemplateSecrets` is pretty nice, they're not just confined to `FieldExtensions` as the router will automatically take the `backstageToken` and make that a secret so I think something around `TemplateSecrets` makes the most sense.

**Review by freben — COMMENTED** [review]
Cool stuff!

File: plugins/scaffolder/src/components/fields/RepoUrlPicker/RepoUrlPicker.tsx [hunk 7]
```diff
@@ -66,6 +81,36 @@ export const RepoUrlPicker = (
...
+        !(state.host && state.owner && state.repoName)
+      ) {
+        return;
+      }
+
+      // user has requested that we use the users credentials
+      // so lets grab them using the scmAuthApi and pass through
+      // any additional scopes from the ui:options
+      const { token } = await scmAuthApi.getCredentials({
+        url: `https://${state.host}/${state.owner}/${state.repoName}`,
```
[the remark this answers is not in the document]
Comment (benjdlambert) [reply plugins/scaffolder/src/components/fields/RepoUrlPicker/RepoUrlPicker.tsx:99]: yeah I mean it's gonna be impossible for us to implement everything that works as we don't have the resources to be able to test everything so I'm kinda hoping that this works for most, and if there's something that needs fixing we can get a contribution for it to make it work 🙏

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts [hunk 8]
```diff
@@ -53,6 +53,7 @@ type TemplateContext = {
   steps: {
     [stepName: string]: { output: { [outputName: string]: JsonValue } };
   };
+  secrets?: Record<string, string>;
```
[the remark this answers is not in the document]
Comment (benjdlambert) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:56]: :joy: the reason I did this is cos in a previous PR we chatted about replacing `JSONObject` with just `Record<string, string>`.  I think let's keep it simple for now 👍

**Review by jhaals — APPROVED** [review]

**Comment by benjdlambert** [comment]
OK - Have a little bit of a better idea here rather than the `useDebounce` around requiring user credentials before being able to navigate to the next step of the form. Think we can do this in a nice way actually without needing #7868 for now.

Will need to implement the `optional` property in `SCMAuth` first however, but hopefully that can be pretty quick side PR>

**Comment by NishkarshRaj** [comment]
@benjdlambert awesome work! This would be of great help 🥳

**Comment by benjdlambert** [comment]
Merged this to get this in and out for testing to people. WIll be looking to improve the UX of asking for credentials from the user by displaying a Login button if they are not logged in, rather than hitting the user with a popup mid typing in the form.

**Pull request merged**